apis/python/src/tiledbsoma/io/shaping.py

```diff
@@ -0,0 +1,546 @@
+        ms_name=None,
+        coll_name=None,
+        verbose=verbose,
+        check_only=check_only,
+        context=context,
+        output_handle=output_handle,
+    )
+    ok = _treewalk(
+        uri,
+        visitor=_leaf_visitor_upgrade,
+        args=args,
+    )
+    return ok
+
+
+def resize_experiment(
+    uri: str,
+    *,
+    nobs: int,
+    nvars: Dict[str, int],
+    verbose: bool = False,
+    check_only: bool = False,
+    context: Optional[tiledbsoma.SOMATileDBContext] = None,
+    output_handle: Printable = cast(Printable, sys.stdout),
+) -> bool:
+    """For each dataframe contained within the SOMA ``Experiment`` pointed to by
+    the given URI, resizes the ``domain`` for the ``soma_joinid index column
+    (if it is an indexed column) to match the desired new value.
+
+    The desired new value may be the same size as at present, or bigger, but not
+    exceeding ``maxdomain`` for the ``soma_joinid`` index column.
+
+    For each N-D array contained within the SOMA ``Experiment``, resizes the ``shape``
+    to match the desired values.
+
+    * For ``X`` arrays, the resize is to new ``nobs`` x the measurement's new ``nvar``.
+    * For ``obsm`` arrays, the resize is to new ``nobs`` x the array's existing ``soma_dim_1`` shape.
+    * For ``obsp`` arrays, the resize is to new ``nobs`` x that same ``nobs``.
+    * For ``varm`` arrays, the resize is to new ``nvar`` x the array's existing ``soma_dim_1`` shape.
+    * For ``varp`` arrays, the resize is to new ``nvar`` x that same ``nvar``.
+
+    In all cases, the desired new ``shape`` value may be the same size on the
+    given dimension, or bigger, but not exceeding ``maxshape`` for the given
+    dimension.
+
+    If any array has not been upgraded, then its resize will fail.
+
+    Args:
+        uri: The URI of a SOMA :class:`Experiment``.
+        nobs: The desired new shape of the experiment's ``obs`` dataframe.
+        nvars: The desired new shapes of the experiment's ``var`` dataframes.
+            This should be a dict from measurement name to shape, e.g.
+            ``{"RNA": 10000, "raw": 20000}``.
+        verbose: If ``True``, produce per-array output as the upgrade runs.
+        check_only: If ``True``,  don't apply the upgrades, but show what would
+            be attempted, and show why each one would fail.
+        context: Optional :class:`SOMATileDBContext``.
+    """
+    args: SizingArgs = dict(
+        nobs=nobs,
+        nvars=nvars,
+        ms_name=None,
+        coll_name=None,
+        verbose=verbose,
+        check_only=check_only,
+        context=context,
+        output_handle=output_handle,
+    )
+
+    # Do an early check on the nvars keys vs the experiment's
+    # measurent names. This isn't a can-do status for the experiment;
+    # it's a failure of the user's arguments.
+    with tiledbsoma.Experiment.open(uri) as exp:
+        arg_keys = sorted(list(set(nvars.keys())))
+        ms_keys = sorted(list(set(exp.ms.keys())))
```

Comment:
Dictionary keys are already be unique, so I don't think the call to `set` is necessary.

```suggestion
        arg_keys = sorted(nvars.keys())
        ms_keys = sorted(exp.ms.keys())
```